packages/babel-parser/src/plugins/jsx/index.js

```diff
@@ -625,22 +614,35 @@ export default (superClass: Class<Parser>): Class<Parser> =>
     }
 
     updateContext(prevType: TokenType): void {
-      if (this.match(tt.braceL)) {
-        const curContext = this.curContext();
+      super.updateContext(prevType);
+      const { context, type } = this.state;
+      if (type === tt.braceL) {
+        const curContext = context[context.length - 1];
         if (curContext === tc.j_oTag) {
-          this.state.context.push(tc.braceExpression);
+          context.push(tc.brace);
         } else if (curContext === tc.j_expr) {
-          this.state.context.push(tc.templateQuasi);
-        } else {
-          super.updateContext(prevType);
+          context.push(tc.templateQuasi);
         }
         this.state.exprAllowed = true;
-      } else if (this.match(tt.slash) && prevType === tt.jsxTagStart) {
-        this.state.context.length -= 2; // do not consider JSX expr -> JSX open tag -> ... anymore
-        this.state.context.push(tc.j_cTag); // reconsider as closing tag context
+      } else if (type === tt.slash && prevType === tt.jsxTagStart) {
+        context.length -= 2; // do not consider JSX expr -> JSX open tag -> ... anymore
+        context.push(tc.j_cTag); // reconsider as closing tag context
+        this.state.exprAllowed = false;
+      } else if (type === tt.jsxTagEnd) {
+        const out = context.pop();
+        if ((out === tc.j_oTag && prevType === tt.slash) || out === tc.j_cTag) {
+          context.pop();
```

Comment:
What are we popping out here? `j_expr`?

Reply:
If `out` is `j_oTag` (`<name></name>`), `context.pop` is `j_expr`

If `out` is `j_cTag` (`<name />`), `context.pop` is the context before `j_cTag`, which is also the one before `j_expr` since we reinterpret `j_expr, j_oTag` as `j_cTag` when we see slash following `jsxTagStart`.

The logic here is was `tt.jsxTagEnd.updateContext`, I move it here so we can simplify the `updateContext` interface.